lib/puppet_x/templates/invoke_ps_command.erb

```diff
@@ -151,38 +347,58 @@ try
 
   $ps = [System.Management.Automation.PowerShell]::Create()
   $ps.Runspace = $runspace
-
   [Void]$ps.AddScript($ourFunctions)
   $ps.Invoke()
 
+
   if(!$environmentVariables){
+    $ps = [System.Management.Automation.PowerShell]::Create()
+    $ps.Runspace = $runspace
     $environmentVariables = $ps.AddCommand('Get-ProcessEnvironmentVariables').Invoke()
   }
 
   if($PSVersionTable.PSVersion -le [Version]'2.0'){
     if(!$psVariables){
       $psVariables = $ps.AddScript('Get-Variable').Invoke()
     }
+
+    $ps = [System.Management.Automation.PowerShell]::Create()
+    $ps.Runspace = $runspace
     [void]$ps.AddScript('Get-Variable -Scope Global | Remove-Variable -Force -ErrorAction SilentlyContinue -WarningAction SilentlyContinue')
     $ps.Invoke()
+
+    $ps = [System.Management.Automation.PowerShell]::Create()
+    $ps.Runspace = $runspace
     [void]$ps.AddCommand('Reset-ProcessPowerShellVariables').AddParameter('psVariables', $psVariables)
     $ps.Invoke()
   }
 
+  $ps = [System.Management.Automation.PowerShell]::Create()
+  $ps.Runspace = $runspace
   [Void]$ps.AddCommand('Reset-ProcessEnvironmentVariables').AddParameter('processVars', $environmentVariables)
   $ps.Invoke()
 
+  $ps = [System.Management.Automation.PowerShell]::Create()
+  $ps.Runspace = $runspace
```

Comment:
I could get Verbose, Debug and Progress streams with our current code, but not the Error stream (see our ticket for explanation). After a long advanced debugging session (read: comment part of code, run, then uncomment, repeat) I found that if we didn't create a new PowerShell object after the call to Reset-ProcessEnvironmentVariables, I would loose the formatting for any Error object and instead just get the message, not the fully formatted string.

Last time i used C# and a PowerShell host, I used a new PowerShell.Create object for each command, but can't point you to a specific documentation site stating the need. The Runspace object is where the state is, the PowerShell object represents the pipeline part, so this makes sense to me. I wanted to make a helper function to reduce the double lines everywhere, but saw the extra function as potentially more confusing down the line.